core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java

```diff
@@ -0,0 +1,171 @@
+package ai.timefold.solver.core.impl.solver.termination;
+
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+import ai.timefold.solver.core.api.score.Score;
+import ai.timefold.solver.core.impl.phase.scope.AbstractPhaseScope;
+import ai.timefold.solver.core.impl.phase.scope.AbstractStepScope;
+import ai.timefold.solver.core.impl.solver.scope.SolverScope;
+import ai.timefold.solver.core.impl.solver.thread.ChildThreadType;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+public final class AdaptiveTermination<Solution_, Score_ extends Score<Score_>> implements Termination<Solution_> {
+    private final long gracePeriodMillis;
+    private final double minimumImprovementRatio;
+
+    boolean isGracePeriodActive;
+    private long gracePeriodStartTimeMillis;
+    private LevelScoreDiff gracePeriodScoreDiff;
+
+    private final NavigableMap<Long, Score_> scoresByTime;
+
+    public AdaptiveTermination(long gracePeriodMillis, double minimumImprovementRatio) {
+        this.gracePeriodMillis = gracePeriodMillis;
+        this.minimumImprovementRatio = minimumImprovementRatio;
+        this.scoresByTime = new TreeMap<>();
+    }
+
+    private record LevelScoreDiff(boolean harderScoreChanged, double softestScoreDiff) {
```

Comment:
Typically, I would be all for a carrier type like this. In this case, and on the hot path, I'd rather store the two variables individually and avoid allocating the instance of the record. 

In fact, this is the only thing that actually made a measurable impact in my experiments - `LevelScoreDiff` was ~10th most allocated object on the heap.